### DISCUSSION TITLE
How to further aggregate already aggregated data in a materialized view?

### DISCUSSION THREAD

**Discussion by pibi** [ClickHouse/ClickHouse#59174] [Q&A]
For a project, I need to aggregate an already aggregated column in a final table. The original use case is quite complex, but I managed to simplify it for this discussion. 
Basically, Let's say I have a table that stores every login from a given id, at a specific date and with a timestamp. I need to know the total number of logins by id since the start of the day to a given minute. For the sake of performances I need that result stored in a table. 

I thought of different strategies so far, but the best one I come up is by using a materialized view: 

```sql
CREATE MATERIALIZED VIEW counter_mv TO counter
(
    `date` Date,
    `id` LowCardinality(String),
    `dt` DateTime('America/New_York'),
    `n` UInt64,
    `partial` SimpleAggregateFunction(sum, UInt64)
) AS
SELECT
    date, id,
    toStartOfInterval(t,toIntervalMinute(1)) AS dt,    
    count(*) AS n,
    sum(n) OVER (PARTITION BY date, id ORDER BY dt ASC) AS partial
FROM logins
GROUP BY
    date, id, dt
```
The query works fine when used in a standalone insert.  But,  when inserting data in `logins`  the `n` and `partial` are equal.  I understand this is because the view works only over the current data batch. Is that right? If this is the case, is there any way to further aggregate the logins to get the current total sum of logins for that id at that minute? Any other strategy?

 thanks  a lot for your support

**Comment by den-crane** [comment]
In general materialized views with windows functions `sum(n) OVER (..) AS partial` not possible. 

https://den-crane.github.io/Everything_you_should_know_about_materialized_views_commented.pdf 

https://youtu.be/QUigKP7iy7Y

https://www.youtube.com/watch?v=QDAJTKZT8y4

but you should use `SummingMergeTree order by (date, id, dt)` in TO table, it will aggregate data.

Also check https://clickhouse.com/docs/en/sql-reference/statements/create/view#refreshable-materialized-view  Refreshable Materialized View can use windows functions `sum(n) OVER (..) AS partial`.

**Reply by pibi** [reply]
Thanks a lot for your pdf. I read it, and now I'm wondering if  `count(*) AS n,`  would really work as intended. Per your **Double grouping** this will only work on the current inserts. Right? I can fix that one using `SummingMergeTree order by (date, id, dt)` in the TO table, as you suggested. But I don't know how to create the `SummingMergeTree`  destination for the `partial` column. 
Refreshable Materialized View looks great, but I cannot recreate the whole table just for the today data. The source is huge, so this will be very inefficient.

I also considered the [Window view](https://clickhouse.com/docs/en/sql-reference/statements/create/view#window-view-experimental) but they work looking forward, while I need to look from the beginning of the day, in the past.

**Comment by den-crane** [comment]
more examples

https://gist.github.com/search?q=user%3Aden-crane+summingmergetree
https://gist.github.com/search?q=user%3Aden-crane+aggregatingmergetree

**Comment by pibi** [comment]
I also tried to define `partial` as a `MATERIALIZED` column in the TO table:
```sql
partial UInt64 MATERIALIZED (sum(n) OVER (PARTITION BY date, id ORDER BY dt ASC))
```
but it seems window function in a MATERIALIZED column is unsupported

**Comment by pibi** [comment]
I found this [issue](https://github.com/ClickHouse/ClickHouse/issues/54818) that looks related to my use-case. It seems there are limitations in every direction about getting the destination table  to compute the cumulative sum at that point. I can't even do that from an `ALTER TABLE ... UPDATE` using a subquery or a join table. Everything just works in a pure select, but it falls apart when doing the same to store that result. 
Even by using `SummingMergeTree order by (date, id, dt)` as suggested it would not work. It still needs to know the value of the previous row. @den-crane It's like computing the `eventsByDay` table in your [example](https://gist.github.com/den-crane/85b67e6573a837ffbcd36c601be66ecd) with an additional column that sums the events in that group since the start of the month.

I'm not really sure how the fix by @amosbird could work in this scenario.

**Comment by den-crane** [comment]
@pibi  https://clickhouse.com/docs/en/sql-reference/statements/create/view#refreshable-materialized-view Refreshable Materialized View can use windows functions sum(n) OVER (..) AS partial.

**Reply by pibi** [reply]
@den-crane Refreshable Materialized View looks great, but I cannot recreate the whole table every time just for the today data. The source is huge, so this will be very inefficient. Am I missing something?

**Reply by den-crane** [reply]
You missing the fact that this is the only way to calculate cumulative sum.

**Reply by pibi** [reply]
hmmm, so, in the original form I cannot use it directly. Too much data in the source table. It will take days to recreate the same results,and a bunch of new rows at the most recent dates.  Maybe I can use it to create data for the most recent day only. But how can I store the computed values in the final column then?

**Reply by UnamedRus** [reply]
BTW,
You can calculate sum per day in Materialized View.
And calculate cumulative part in query time. 
Usually preaggregated only first step give good enough perf boost.
